server/core/src/main/java/dev/slimevr/tracking/processor/Constraint.kt

```diff
@@ -0,0 +1,183 @@
+
+				val magnitude = if (distToMin < distToMax) magnitudeMin else magnitudeMax
+				val magnitudeSqr = abs(if (distToMin < distToMax) magnitudeSqrMin else magnitudeSqrMax)
+				vector = vector.unit() * -magnitude
+
+				rot = Quaternion(sqrt(1.0f - magnitudeSqr), vector)
+			}
+
+			return rot.unit()
+		}
+
+		// Constraint function for TwistSwingConstraint
+		private val twistSwingConstraint: ConstraintFunction =
+			{ rotation: Quaternion, thisBone: Bone, swingRad: Float, twistRad: Float, _: Float ->
+				if (thisBone.parent == null) {
+					rotation
+				} else {
+					val parent = thisBone.parent!!
+					val localRotationOffset = parent.rotationOffset.inv() * thisBone.rotationOffset
+					val rotationLocal =
+						(parent.getGlobalRotation() * localRotationOffset).inv() * rotation
+					var (swingQ, twistQ) = decompose(rotationLocal, Vector3.NEG_Y)
+
+					swingQ = constrain(swingQ, swingRad)
+					twistQ = constrain(twistQ, twistRad)
+
+					(parent.getGlobalRotation() * localRotationOffset * (swingQ * twistQ)).unit()
+				}
+			}
+
+		// Constraint function for a hinge constraint with min and max angles
+		private val hingeConstraint: ConstraintFunction =
+			{ rotation: Quaternion, thisBone: Bone, min: Float, max: Float, _: Float ->
+				if (thisBone.parent == null) {
+					rotation
+				} else {
+					val parent = thisBone.parent!!
+					val localRotationOffset = parent.rotationOffset.inv() * thisBone.rotationOffset
+					val rotationLocal =
+						(parent.getGlobalRotation() * localRotationOffset).inv() * rotation
+
+					var (_, hingeAxisRot) = decompose(rotationLocal, Vector3.NEG_X)
+
+					hingeAxisRot = constrain(hingeAxisRot, min, max, Vector3.NEG_X)
+
+					(parent.getGlobalRotation() * localRotationOffset * hingeAxisRot).unit()
+				}
+			}
+
+		// Constraint function for a hinge constraint with min and max angles that allows nonHingeDeviation
+		// rotation on all axis but the hinge
+		private val looseHingeConstraint: ConstraintFunction =
+			{ rotation: Quaternion, thisBone: Bone, min: Float, max: Float, nonHingeDeviation: Float ->
+				if (thisBone.parent == null) {
+					rotation
+				} else {
+					val parent = thisBone.parent!!
+					val localRotationOffset = parent.rotationOffset.inv() * thisBone.rotationOffset
+					val rotationLocal =
+						(parent.getGlobalRotation() * localRotationOffset).inv() * rotation
+
+					var (nonHingeRot, hingeAxisRot) = decompose(rotationLocal, Vector3.NEG_X)
+
+					hingeAxisRot = constrain(hingeAxisRot, min, max, Vector3.NEG_X)
+					nonHingeRot = constrain(nonHingeRot, nonHingeDeviation)
+
+					(parent.getGlobalRotation() * localRotationOffset * (nonHingeRot * hingeAxisRot)).unit()
+				}
+			}
+
+		// Constraint function for CompleteConstraint
+		private val completeConstraint: ConstraintFunction = { _: Quaternion, thisBone: Bone, _: Float, _: Float, _: Float ->
+			thisBone.getGlobalRotation()
```

Comment:
Should we not return the initial rotation provided? I think using the bone is beyond the necessary scope.

Reply:
See my other comment on the input rotation not being the same as the bone's global rotation when constraints are being used in an IK solver. I did change to returning the input rotation but the completeConstraint function should never be called due to it being handled before the constraint function is called due to its need for the bone.